packages/vscode-extension/lib/network/networkRequestParsers.ts

```diff
@@ -86,23 +110,84 @@ const TRUNCATED_LENGTH = 1000; // 1000 characters
  * @param responseBody The string body to inspect and potentially truncate.
  * @returns An InternalResponseBodyData with `body`, `wasTruncated` and `dataSize`.
  */
-function truncateResponseBody(responseBody: string | undefined): InternalResponseBodyData {
+function parseResponseBodyData(
+  responseBody: string | undefined,
+  base64Encoded: boolean = false,
+  type: ResponseBodyDataType = ResponseBodyDataType.Other
+): InternalResponseBodyData {
   if (!responseBody) {
-    return { body: undefined, wasTruncated: false, dataSize: 0 };
+    return DEFAULT_RESPONSE_BODY_DATA;
   }
 
   const dataSize = new Blob([responseBody]).size;
 
   if (dataSize > MAX_BODY_SIZE) {
     const slicedBody = responseBody.slice(0, TRUNCATED_LENGTH);
     return {
-      body: `${slicedBody}...`,
+      body: slicedBody,
       wasTruncated: true,
+      base64Encoded,
+      type,
       dataSize: new Blob([slicedBody]).size,
     };
   }
 
-  return { body: responseBody, wasTruncated: false, dataSize };
+  return { body: responseBody, base64Encoded, wasTruncated: false, type, dataSize };
+}
+
+function handleReadError(error: unknown): InternalResponseBodyData {
+  console.warn("Failed to read response body content:", error);
+  return parseResponseBodyData(undefined);
+}
+
+/**
+ * Image and octet-stream handling.
+ */
+function readBlobAsBase64(blob: Blob, isImage?: boolean): Promise<InternalResponseBodyData> {
+  return new Promise((resolve) => {
+    const reader = new FileReader();
+
+    reader.onload = () => {
+      let textResult: string | undefined;
+      if (reader.result instanceof ArrayBuffer) {
+        const uint8Array = new Uint8Array(reader.result);
+        textResult = dataToBase64(uint8Array);
+      } else if (typeof reader.result === "string") {
+        textResult = reader.result;
+      }
+
+      const type = isImage ? ResponseBodyDataType.Image : ResponseBodyDataType.Other;
+
+      resolve(parseResponseBodyData(textResult, true, type));
+    };
+
+    reader.onerror = (error) => {
+      resolve(handleReadError(error));
+    };
+
+    reader.readAsArrayBuffer(blob);
+  });
+}
+
+/**
+ * Text and parsable application types handling.
+ */
+function readBlobAsText(blob: Blob): Promise<InternalResponseBodyData> {
+  return new Promise((resolve) => {
+    const reader = new FileReader();
+
+    reader.onload = () => {
+      const result = reader.result;
+      const textResult = typeof result === "string" ? result : undefined;
+      resolve(parseResponseBodyData(textResult, false, ResponseBodyDataType.Other));
+    };
+
+    reader.onerror = (error) => {
+      resolve(handleReadError(error));
+    };
```

Comment:
```suggestion
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.onload = () => {
      const result = reader.result;
      const textResult = typeof result === "string" ? result : undefined;
      resolve(parseResponseBodyData(textResult, false, ResponseBodyDataType.Other));
    };

    reader.onerror = (error) => {
      reject(error);
    };
```
(and likewise for `readBlobAsBase64`) would let you handle all your errors in one `catch`, and feels more natural.

Reply:
Changed in 2b9ba016. The promises are now rejected and handled inside `readResponseText`.

```ts
async function readResponseText(
  xhr: XMLHttpRequest
): Promise<InternalResponseBodyData | undefined> {
  try {
    // @ts-ignore - RN-specific property // @ts-ignore - RN-specific property
    if (!xhr || !xhr._cachedResponse) {
      // if response was not accessed, it's not cached and we don't want to read it
      // here to avoid potential side effects
      return undefined;
    }

    const responseType = xhr.responseType;

    if (responseType === "" || responseType === "text") {
      return parseResponseBodyData(xhr.responseText);
    }

    if (responseType === "blob") {
      const contentType = getContentTypeHeader(xhr);
      const isTextType = contentType.startsWith("text/");
      const isParsableApplicationType = Array.from(PARSABLE_APPLICATION_CONTENT_TYPES).some(
        (type) => contentType.startsWith(type)
      );
      const isImageType = contentType.startsWith("image/");
      const isOctetStream = contentType === "application/octet-stream";

      let promise: Promise<InternalResponseBodyData> | undefined = undefined;

      if (isImageType || isOctetStream) {
        promise = readBlobAsBase64(xhr.response, isImageType);
      }

      if (isTextType || isParsableApplicationType) {
        promise = readBlobAsText(xhr.response);
      }

      return promise?.catch((error) => handleReadResponseError(error));
    }

    // don't read binary data
    return undefined;
  } catch (error) {
    // in case synchronous errors occur
    return handleReadResponseError(error);
  }
}
```